res/controllers/Arturia-KeyLab-Mk1-scripts.js

```diff
@@ -0,0 +1,759 @@
+                setLED(0x1F, !value);
+                if (!noWrite) {
+                    const currentDeckNum = this.activeDeck.currentDeckNum();
+                    writeDisplay(`Deck ${currentDeckNum}`);
+                }
+            };
+        }
+        constructor(id) {
+            super({});
+
+            this.id = id;
+            this.activeDeck = new Deck([1, 2, 3, 4], 1);
+            this.deck1Button = new DeckButton([1, 3], this.activeDeck, {
+                group: "[Channel1]",
+                midi: [0xB0, 0x76, 0x1E], // "Sound"
+                output: this.setDeckOutput(),
+            });
+            this.deck2Button = new DeckButton([2, 4], this.activeDeck, {
+                group: "[Channel2]",
+                midi: [0xB0, 0x77, 0x1F], // "Multi"
+                output: this.setDeckOutput(),
+            });
+
+            // Deck parameter selector.
+            this.paramEncoder = new components.Encoder({
+                group: "[Channel1]",
+                midi: [0xB0, 0x70], // Param encoder
+                selectedParam: 0,
+                params: [...menu],
+                activeDeck: this.activeDeck,
+                input: function(_channel, _control, value, _status, _group) {
+                    if (value === 0x3F) {
+                        this.selectedParam = (((this.selectedParam - 1) % this.params.length) + this.params.length)
+                            % this.params.length;
+                    } else if (value === 0x41) {
+                        this.selectedParam = (((this.selectedParam + 1) % this.params.length) + this.params.length)
+                            % this.params.length;
+                    }
+                    const param = this.params[this.selectedParam];
+                    if (param.group) {
+                        this.activeDeck.valueEncoder.group = param.group;
+                        this.activeDeck.valueButton.group = param.group;
+                    } else {
+                        this.activeDeck.valueEncoder.group = this.activeDeck.currentDeck;
+                        this.activeDeck.valueButton.group = this.activeDeck.currentDeck;
+                    }
+                    this.activeDeck.valueEncoder.name = param.name;
+                    this.activeDeck.valueButton.name = param.name;
+                    this.activeDeck.valueEncoder.inKey = param.inKey;
+                    this.activeDeck.valueEncoder.max = param.max;
+                    this.activeDeck.valueEncoder.min = param.min;
+                    this.activeDeck.valueEncoder.relative = param.relative || false;
+                    this.activeDeck.valueButton.inKey = param.pressInKey;
+                    this.activeDeck.valueButton.type = param.pressType || components.Button.prototype.types.push;
+                    this.activeDeck.valueEncoder.outputValue = param.outputValue;
+                    this.activeDeck.valueButton.outputValue = param.outputValue;
```

Comment:
`Object.assign` may be useful to clean this up.

Reply:
I thought of that, but since the key isn't always the same name as what Mixxx expects (ie. "pressInKey" for the buttons "inKey") that doesn't work. Maybe I should swap the menu over to have separate "button" and "encoder" objects that have the expected key names so I can just copy that whole object instead. Although it still feels odd to not explicitly show what's supported. I'll play with it and see what works. Thanks.

Reply:
Done, this is much cleaner looking now. I also noticed a minor bug where the name wasn't being set so it showed up as "Undefined" on the menu when, so eg. when sorting columns it would say "Undefined: artist" instead of "Sort: artist". I also removed some unnecessary copying, the menu slice is never modified so we can just reference it directly.